glean_parser/metrics.py

```
@@ -0,0 +1,149 @@
+# -*- coding: utf-8 -*-
+
+"""
+Classes for each of the high-level metric types.
+"""
+
+from dataclasses import dataclass, field
+from typing import List, Union
+
+import jsonschema
+
+from . import parser
+
+
+@dataclass
+class Metric:
+    metric_types = {}
+
+    def __init_subclass__(cls, **kwargs):
+        # Create a mapping of all of the subclasses of this class
+        if cls not in Metric.metric_types and hasattr(cls, 'typename'):
+            Metric.metric_types[cls.typename] = cls
+        super().__init_subclass__(**kwargs)
+
+    @classmethod
+    def make_metric(cls, group_name, name, metric_info):
+        """
+        Given a metric_info dictionary from metrics.yaml, return a metric
+        instance.
+        """
+        metric_type = metric_info['type']
+        return cls.metric_types[metric_type](
+            group_name=group_name,
+            name=name,
+            **metric_info
+        )
+
+    def serialize(self):
+        """
+        Serialize the metric back to JSON object model.
+        """
+        d = dict(self.__dict__)
+        del d['name']
+        del d['group_name']
+        return d
+
+    def __post_init__(self):
+        schema = parser._get_metrics_schema()
+        data = {
+            self.group_name: {
+                self.name: self.serialize()
+            }
+        }
+        jsonschema.validate(data, schema)
```

This only reports the first error in the schema. Since [bug 1401612](https://bugzilla.mozilla.org/show_bug.cgi?id=1401612), we switched to a model where parsers lazily fail: they first collect all the problems and then report them all, without failing early. I think this can be achieved with the `jsonschema` with something like:

```
validator_class = jsonschema.validators.validator_for(schema)
validator = validator_class(schema)
for error in sorted(validator.iter_errors(data), key=str):
    print(error.message)
```

---

I agree: this would be good to do.  I'll go back and refactor things to make this possible.

---

I think this is done now.  It was a more major change than I expected, but definitely worth doing.